Content.Server/MachineLinking/System/SignalTimerSystem.cs

```diff
@@ -0,0 +1,165 @@
+                {
+                    Trigger(timer.Owner, timer);
+
+                    if (timer.DoneSound != null)
+                    {
+                        var filter = Filter.Pvs(timer.Owner, entityManager: EntityManager);
+                        _audio.Play(timer.DoneSound, filter, timer.Owner, timer.SoundParams);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        ///     Checks if a UI <paramref name="message"/> is allowed to be sent by the user.
+        /// </summary>
+        /// <param name="uid">The entity that is interacted with.</param>
+        private bool IsMessageValid(EntityUid uid, BoundUserInterfaceMessage message)
+        {
+            if (message.Session.AttachedEntity is not { Valid: true } mob)
+                return false;
+
+            if (!_interaction.InRangeUnobstructed(mob, uid))
+                return false;
+
+            if (!_accessReader.IsAllowed(mob, uid))
+                return false;
+
+            return true;
+        }
+
+        private void OnTextChangedMessage(EntityUid uid, SignalTimerComponent component, SignalTimerTextChangedMessage args)
+        {
+            if (!IsMessageValid(uid, args))
+                return;
+
+            component.Label = args.Text[..Math.Min(5,args.Text.Length)];
+            _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, component.Label);
+        }
+
+        private void OnDelayChangedMessage(EntityUid uid, SignalTimerComponent component, SignalTimerDelayChangedMessage args)
+        {
+            if (!IsMessageValid(uid, args))
+                return;
+
+            component.Delay = args.Delay.TotalSeconds;
+        }
+
+        private void OnTimerStartMessage(EntityUid uid, SignalTimerComponent component, SignalTimerStartMessage args)
+        {
+            if (!IsMessageValid(uid, args))
+                return;
+
+            if (!HasComp<ActiveSignalTimerComponent>(uid))
+            {
+                var activeTimer = EnsureComp<ActiveSignalTimerComponent>(uid);
+                activeTimer.TriggerTime = _gameTiming.CurTime + TimeSpan.FromSeconds(component.Delay);
+                component.User = args.User;
+
+                _appearanceSystem.SetData(uid, TextScreenVisuals.Mode, TextScreenMode.Timer);
+                _appearanceSystem.SetData(uid, TextScreenVisuals.TargetTime, activeTimer.TriggerTime);
+                _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, component.Label);
+
+                _signalSystem.InvokePort(uid, component.StartPort);
+            }
+            else
+            {
+                component.User = args.User;
+                RemComp<ActiveSignalTimerComponent>(uid);
+
+                _appearanceSystem.SetData(uid, TextScreenVisuals.Mode, TextScreenMode.Text);
+                _appearanceSystem.SetData(uid, TextScreenVisuals.ScreenText, component.Label);
+            }
```

Comment:
Should be passing around AppearanceComponent? because SetData is doing a resolve internally each time.